pex/dependency_manager.py

```diff
@@ -0,0 +1,108 @@
+from pex.orderedset import OrderedSet
+from pex.pep_503 import ProjectName
+from pex.pex_builder import PEXBuilder
+from pex.pex_info import PexInfo
+from pex.resolve.resolvers import Installed
+from pex.tracer import TRACER
+from pex.typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from typing import DefaultDict, Iterable, Iterator
+
+    import attr  # vendor:skip
+else:
+    from pex.third_party import attr
+
+
+@attr.s
+class DependencyManager(object):
+    _requirements = attr.ib(factory=OrderedSet)  # type: OrderedSet[Requirement]
+    _distributions = attr.ib(factory=OrderedSet)  # type: OrderedSet[FingerprintedDistribution]
+
+    def add_from_pex(self, pex):
+        # type: (str) -> None
+
+        pex_info = PexInfo.from_pex(pex)
+        self._requirements.update(Requirement.parse(req) for req in pex_info.requirements)
+
+        pex_environment = PEXEnvironment.mount(pex, pex_info=pex_info)
+        self._distributions.update(pex_environment.iter_distributions())
+
+    def add_from_installed(self, installed):
+        # type: (Installed) -> None
+
+        for installed_dist in installed.installed_distributions:
+            self._requirements.update(installed_dist.direct_requirements)
+            self._distributions.add(installed_dist.fingerprinted_distribution)
+
+    def configure(
+        self,
+        pex_builder,  # type: PEXBuilder
+        excluded=(),  # type: Iterable[str]
+    ):
+        # type: (...) -> None
+
+        exclude_configuration = ExcludeConfiguration.create(excluded)
+        exclude_configuration.configure(pex_builder.info)
+
+        dists_by_project_name = defaultdict(
+            OrderedSet
+        )  # type: DefaultDict[ProjectName, OrderedSet[FingerprintedDistribution]]
+        for dist in self._distributions:
+            dists_by_project_name[dist.distribution.metadata.project_name].add(dist)
+
+        def iter_non_excluded_distributions(requirements):
+            # type: (Iterable[Requirement]) -> Iterator[FingerprintedDistribution]
+            for req in requirements:
+                candidate_dists = dists_by_project_name[req.project_name]
+                for candidate_dist in tuple(candidate_dists):
+                    if candidate_dist.distribution not in req:
+                        continue
+                    candidate_dists.discard(candidate_dist)
+
+                    excluded_by = exclude_configuration.excluded_by(candidate_dist.distribution)
+                    if excluded_by:
+                        excludes = " and ".join(map(str, excluded_by))
+                        TRACER.log(
+                            "Skipping adding {candidate}: excluded by {excludes}".format(
+                                candidate=candidate_dist.distribution, excludes=excludes
+                            )
+                        )
+                        for root_req in self._requirements:
+                            if candidate_dist.distribution in root_req:
```

Comment:
With a significant number of excludes, these nested loops are potentially O(n**2) (or worse), is that a concern?

Reply:
I was concerned about the outer double loop and so constructed `dists_by_project_name` to limit the outer inner; so that's the best it can be afaict using standard data structures.

Here there is an additional single loop (that scales with root input requirements as opposed to excludes) that could only be sped up by either killing the warning, or keeping it and similarly using a pre-built multimap keyed on project_name. The biggest input requirement sets I've seen are on the order of 100. At full exclusion, that's 10k checks. I'd be surprised if that amounted to more than ~500ms total, but I've gone ahead and added the multimap.